tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -781,6 +781,161 @@ int main(int argc, char **argv)
+                &s2n_ecdhe_rsa_with_chacha20_poly1305_sha256,   /* 0xCC,0xA8 */
+            };
+                
+            const struct s2n_cipher_preferences cipher_preferences_test_arbitrary_equal_preferences_tls13 = {
+                .count = s2n_array_len(s2n_test_arbitrary_equal_preference),
+                .suites = s2n_test_arbitrary_equal_preference,
+            };
+            const struct s2n_security_policy test_arbitrary_equal_preference = {
+                .minimum_protocol_version = S2N_SSLv3,
+                .cipher_preferences = &cipher_preferences_test_arbitrary_equal_preferences_tls13,
+                .kem_preferences = &kem_preferences_null,
+                .signature_preferences = &s2n_signature_preferences_20201021,
+                .ecc_preferences = &s2n_ecc_preferences_test_all,
+            };
+
+            /* Test equal-preferenced tls13 cipher suites */
+            {
+                EXPECT_SUCCESS(s2n_enable_tls13_in_test());
+
+                conn->security_policy_override = &test_all_equal_preference_tls13;
+                conn->kex_params.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+                conn->client_protocol_version = S2N_TLS13;
+                conn->actual_protocol_version = S2N_TLS13;
+                conn->server_protocol_version = S2N_TLS13;
+                
+                uint8_t chacha20_boosted_wire[] = {
+                    TLS_CHACHA20_POLY1305_SHA256,
+                    TLS_AES_256_GCM_SHA384,
+                    TLS_AES_128_GCM_SHA256,
+                };
+
+                uint8_t count = sizeof(chacha20_boosted_wire) / S2N_TLS_CIPHER_SUITE_LEN;
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(conn, chacha20_boosted_wire, count));
+
+                if (s2n_chacha20_poly1305.is_available()) {
+                    EXPECT_EQUAL(conn->secure->cipher_suite, &s2n_tls13_chacha20_poly1305_sha256);
+                } else {
+                    /* Security policy has AES 256 and 128 as equal preference */
+                    EXPECT_EQUAL(conn->secure->cipher_suite, &s2n_tls13_aes_256_gcm_sha384);
+                }
+
+                uint8_t aes256_boosted_wire[] = {
+                    TLS_AES_256_GCM_SHA384,
+                    TLS_CHACHA20_POLY1305_SHA256,
+                    TLS_AES_128_GCM_SHA256,
+                };
+
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(conn, aes256_boosted_wire, count));
+                EXPECT_EQUAL(conn->secure->cipher_suite, &s2n_tls13_aes_256_gcm_sha384);
+
+                uint8_t aes128_boosted_wire[] = {
+                    TLS_AES_128_GCM_SHA256,
+                    TLS_AES_256_GCM_SHA384,
+                    TLS_CHACHA20_POLY1305_SHA256,
+                };
+
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(conn, aes128_boosted_wire, count));
+                EXPECT_EQUAL(conn->secure->cipher_suite, &s2n_tls13_aes_128_gcm_sha256);
```

Comment:
I notice you're not testing
```
               uint8_t aes128_boosted_wire[] = {
                    TLS_AES_256_GCM_SHA384,
                    TLS_AES_128_GCM_SHA256,
                    TLS_CHACHA20_POLY1305_SHA256,
                };
```
In that case, with your security policy, we'd choose TLS_AES_256_GCM_SHA384 over TLS_AES_128_GCM_SHA256. That's probably not correct behavior. The server should be choosing between AES cipher suites. We only want to give the client control of chacha20 vs aes.

So in production security policies, we wouldn't want TLS_AES_256_GCM_SHA384 and TLS_AES_128_GCM_SHA256 to be equal priority. Like:
```
            static struct s2n_cipher_suite *s2n_all_tls13_cipher_suites_equal_preference[] = {
                &s2n_equal_preference_group_start,              /* start group */
                &s2n_tls13_chacha20_poly1305_sha256,            /* 0x13,0x03 */
                &s2n_tls13_aes_128_gcm_sha256,                  /* 0x13,0x01 */
                &s2n_equal_preference_group_end,                /* end group */
                &s2n_tls13_aes_256_gcm_sha384,                  /* 0x13,0x02 */
            };
```

But does that still behave in a sane way that meets customer "chacha20 boosting" use cases? What if the client sends [TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_AES_128_GCM_SHA256]?

Reply:
> In that case, with your security policy, we'd choose TLS_AES_256_GCM_SHA384 over TLS_AES_128_GCM_SHA256. That's probably not correct behavior.

What do you mean by this? This is the correct behaviour for the server's security policy (eg. all 3 ciphers in one equal preference group). So if my test case used a client wire where aes_256 has precedence over aes_128 then aes_256 would get negotiated.

In your last example, the server would negotiate `&s2n_tls13_chacha20_poly1305_sha256`. The algorithm iterates over server cipher suite preferences; since chacha20 and aes_128 are the most-preferred by the server, we turn to the client as a tie-breaker. Because the client prefers chacha20 over aes_128, the server chooses chacha20.

Reply:
I'm not questioning if this test reflects the behavior you implemented. I'm questioning whether this test (and therefore your code) reflects the behavior we want for chacha20 boosting.

Reply:
Isn't this the behaviour we want? If the s2n server were to use a security policy with aes128, aes256 and chacha20 in the same group then we would get chacha20 boosting if the client indicates it. 

The example you gave is also chacha20 boosting; the server was able to negotiate the chacha20 cipher over aes_128 because the client indicated it.

Reply:
Summary from the offline discussion:
* The current security policy lets the client choose whatever AES cipher suite they want, regardless of server preferences. Is that the desired behavior if we're just trying to enable chacha20 boosting?
* The modified security policy in my comment will boost chacha20 even when it's not the first cipher suite offered by the client, as long as it's more preferred than the server's most preferred AES cipher suite. Is that the desired behavior if we're trying to enable chacha20 boosting?

Reply:
An example that might help, since TLS1.3 only has three cipher suites right now. Imagine it had more:
```
static struct s2n_cipher_suite *default[] = {
     &s2n_tls13_aes_128_gcm_sha256,
     &s2n_tls13_aes_256_gcm_sha384, 
     &s2n_tls13_better_than_chacha20,
     &s2n_tls13_chacha20_poly1305_sha256,
     &s2n_tls13_worse_than_chacha20,
};
```

How would you do chacha20 boosting? You either put all the options in one equal preference group, which seems significantly worse now that there are more options and they include s2n_tls13_worse_than_chacha20. Or you do:
```
static struct s2n_cipher_suite *default_with_chacha20_boosted[] = {
     &s2n_equal_preference_group_start,
     &s2n_tls13_chacha20_poly1305_sha256,
     &s2n_tls13_aes_128_gcm_sha256, 
     &s2n_equal_preference_group_end, 

     &s2n_tls13_aes_256_gcm_sha384, 
     &s2n_tls13_better_than_chacha20, 
     &s2n_tls13_worse_than_chacha20,
};
```

But now if the client sends: [s2n_tls13_better_than_chacha20, s2n_tls13_worse_than_chacha20, s2n_tls13_aes_256_gcm_sha384, s2n_tls13_chacha20_poly1305_sha256, s2n_tls13_aes_128_gcm_sha256], the server would pick chacha20. That seems wrong-- it's usually the server's second to last choice, and it's also the client's second to last choice!

We could assume that there will never be more TLS1.3 cipher suites we prefer more than chacha20, or never one that we prefer less, but those don't seem like good / flexible assumptions.